### PR TITLE
Add dimensions to megacomplex and dataset_descriptor

### PR DESCRIPTION
**Pull request by joernweissenborn**
### Change summary

- added `dimension` property and `index_dependent` method to `Megacomplex`
- added `index_dependent`, `get_model_dimension` and `get_global_dimension` to `DatasetDescriptor`
- deprecated `Model.index_dependent`, `Model.global_dimension` and `Model.index_dependent`

### Checklist

- [ ] ✔️ Passing the tests (mandatory for all PR's)
- [-] 👌 Closes issue (mandatory for ✨ feature and 🩹 bug fix PR's)
- [x] 🧪 Adds new tests for the feature (mandatory for ✨ feature and 🩹 bug fix PR's)
- [x] 📚 Adds documentation of the feature

### Closes issues

closes #XXXX

### CODE REVIEW

**Comment by s-weigand** [comment]
Side note:
The `PR Benchmarks` workflow fails because this branch isn't based on the latest `glotaran/pyglotaran@main` so the lookup of `upstream/main^` fails.

**Review by jsnel — CHANGES_REQUESTED** [review]
Rename DatasetDescriptor to DatasetModel (and rename derived classes accordingly) 
Please make the change from dataset_descriptor to dataset_model consistent everywhere, including when calling `model.dataset["name"]` or `model.dataset["name"]`

File: glotaran/model/megacomplex.py [hunk 1]
```diff
@@ -1,12 +1,26 @@
...
 
+if TYPE_CHECKING:
+    from typing import Any
+
+
+def megacomplex(
+    dimension: str,
+    properties: Any | dict[str, dict[str, Any]] = {},
+    attributes: dict[str, dict[str, Any]] = {},
+    dataset_attributes: dict[str, dict[str, Any]] = {},
```
Comment (jsnel) [inline glotaran/model/megacomplex.py:18]: Please refrain from using {} as default argument

File: glotaran/project/scheme.py [hunk 2]
```diff
@@ -36,6 +36,7 @@ class Scheme:
     group_tolerance: float = 0.0
     non_negative_least_squares: bool = False
     maximum_number_function_evaluations: int = None
+    add_svd: bool = True
```
Comment (jsnel) [inline glotaran/project/scheme.py:39]: I'd like this to be False by default, but this requires adapting tests, so maybe in a next PR.

File: glotaran/analysis/problem.py [hunk 3]
```diff
@@ -463,7 +434,7 @@ def create_result_dataset(self, label: str, copy: bool = True) -> xr.Dataset:
         dataset["fitted_data"] = dataset.data - dataset.residual
         return dataset
 
-    def _create_svd(self, name: str, dataset: xr.Dataset):
+    def _create_svd(self, name: str, dataset: xr.Dataset, lsv_dim, rsv_dim):
```
Comment (jsnel) [inline glotaran/analysis/problem.py:437]: Type annotation can be added here

File: glotaran/model/megacomplex.py [hunk 4]
```diff
@@ -1,12 +1,26 @@
...
 import numpy as np
 
 from glotaran.model import DatasetDescriptor
 from glotaran.model import model_attribute
 
+if TYPE_CHECKING:
+    from typing import Any
+
+
+def megacomplex(
```
Comment (jsnel) [inline glotaran/model/megacomplex.py:14]: Please add some documentation for the new megacomplex decorator or the MegaComplex class for that matter; it is an important concept that is at the heart of the current refactoring.

**Comment by s-weigand** [comment]
Before this PR gets merged the deprecations also need the basic tests, so we get failing tests if a deprecation is overdue.
E.g. see
https://github.com/glotaran/pyglotaran/blob/b5e704a1966c9b60168ff31c2f6be4ecc47ceb16/glotaran/deprecation/modules/test/test_model_base_model.py#L1-L10

This is just about executing/accessing something so the deprecation wrapper gets executed.

Note to myself `OverDueDeprecation` should inherit from `BaseException` so it won't be filtered out by `deprecation_warning_on_call_test_helper`.

File: glotaran/analysis/problem.py (hunk 3 above)
Comment (joernweissenborn) [reply glotaran/analysis/problem.py]: fixed

File: glotaran/model/megacomplex.py (hunk 4 above)
Comment (joernweissenborn) [reply glotaran/model/megacomplex.py:14]: Docstring added

**Comment by jsnel** [comment]
> Please rename `DatasetDescriptor` to `DatasetModel` (and rename derived classes and types accordingly), e.g. think of `KineticSpectrumDatasetDescriptor`, `GaussianShapeDecayDatasetDescriptor`, `DecayDatasetDescriptor`, `_T_DatasetDescriptor` , etc
> Please make the change from dataset_descriptor to dataset_model consistent everywhere, including when calling `model.dataset["name"]` or `model.dataset["name"].fill(model,parameters)`

This is to be addressed in a upcoming PR.

**Review by jsnel — APPROVED** [review]
Extensive review by all contributors.

**Comment by jsnel** [comment]
Accepting with issues noted in #717

**Pull request merged**